File: examples/python/async/create_raw_iterator.py
import asyncio
from datetime import timedelta

from BinaryOptionsToolsV2.pocketoption import PocketOptionAsync
from BinaryOptionsToolsV2.validator import Validator


async def main(ssid: str):
    # Use context manager to ensure connection is authenticated and assets are loaded
    async with PocketOptionAsync(ssid) as api:
        # Create a validator for price updates
        validator = Validator.regex(r'\{"price":\d+\.\d+\}')
        # Create an iterator with 5 minute timeout
        stream = await api.create_raw_iterator(
            '42["price/subscribe"]',  # WebSocket subscription message
            validator,
            timeout=timedelta(minutes=5),
        )

        try:
            # Process messages as they arrive
            async for message in stream:
                print(f"Received message: {message}")
        except TimeoutError:
            print("Stream timed out after 5 minutes")
        except Exception as e:
            print(f"Error processing stream: {e}")


if __name__ == "__main__":
    ssid = input("Please enter your ssid: ")
    asyncio.run(main(ssid))

File: examples/javascript/create_raw_iterator.js
// Iterates over every raw response matching a validator.
//
//   node create_raw_iterator.js "<ssid>"

const { PocketOption, Validator } = require("../../nodejs");

async function main(ssid) {
  const api = await PocketOption.create(ssid);

  const validator = Validator.contains('"signals"');

  // Sends the message, then yields matching responses for 10 seconds.
  const stream = await api.createRawIterator('42["signals/subscribe"]', validator, 10_000);

  for await (const message of stream) {
    console.log(`Received: ${message}`);
  }

  console.log("Iterator finished");
  await api.shutdown();
}

const ssid = process.argv[2] || process.env.POCKET_OPTION_SSID;
main(ssid).catch(console.error);
